libpromises/evalfunction.c

```diff
@@ -7777,6 +7780,190 @@ static int JsonPrimitiveComparator(JsonElement const *left_obj,
+        return false;
+    }
+
+    if (JsonGetType(json_child) != JSON_TYPE_STRING)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Function %s(): Bad class expression at key '%s': Expected type string",
+            fn_name, class_expr_key);
+        return false;
+    }
+
+    const char *class_expr = JsonPrimitiveGetAsString(json_child);
+    assert(class_expr != NULL);
+
+    *remove = !IsDefinedClass(ctx, class_expr);
+    return true;
+}
+
+static bool ClassFilterDataArray(
+    EvalContext *ctx,
+    const char *fn_name,
+    const char *data_structure,
+    const char *key_or_index,
+    JsonElement *child,
+    bool *remove)
+{
+    switch (JsonGetType(child))
+    {
+    case JSON_TYPE_ARRAY:
+        if (StringEqual(data_structure, "auto") ||
+            StringEqual(data_structure, "array_of_arrays"))
+        {
+            return ClassFilterDataArrayOfArrays(
+                ctx, fn_name, child, key_or_index, remove);
+        }
+        Log(LOG_LEVEL_VERBOSE,
+            "Function %s(): Expected child element to be of container type array",
+            fn_name);
+        break;
+
+    case JSON_TYPE_OBJECT:
+        if (StringEqual(data_structure, "auto") ||
+            StringEqual(data_structure, "array_of_objects"))
+        {
+            return ClassFilterDataArrayOfObjects(
+                ctx, fn_name, child, key_or_index, remove);
+        }
+        Log(LOG_LEVEL_VERBOSE,
+            "Function %s(): Expected child element to be of container type object",
+            fn_name);
+        break;
+
+    default:
+        Log(LOG_LEVEL_VERBOSE,
+            "Function %s(): Expected child element to be of container type",
+            fn_name);
+        break;
+    }
+
+    return false;
+}
+
+static FnCallResult FnCallClassFilterData(
+    EvalContext *ctx,
+    ARG_UNUSED Policy const *policy,
+    FnCall const *fp,
+    Rlist const *args)
+{
+    assert(ctx != NULL);
+    assert(fp != NULL);
+    assert(args != NULL);
+    assert(args->next != NULL);
+    assert(args->next->next != NULL);
+
+    bool allocated = false;
+    JsonElement *parent = VarNameOrInlineToJson(ctx, fp, args, false, &allocated);
+    if (parent == NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Function %s(): Expected parent element to be of container type array",
+            fp->name);
+        return FnFailure();
+    }
+
+    /* Currently only parent type array is supported */
+    if (JsonGetType(parent) != JSON_TYPE_ARRAY)
+    {
+        JsonDestroyMaybe(parent, allocated);
+        return FnFailure();
+    }
+    assert(allocated); /* Non-primitives are always allocated */
+
+    const char *data_structure = RlistScalarValue(args->next);
+    const char *key_or_index = RlistScalarValue(args->next->next);
+
+    /* Iterate through array backwards so we can avoid having to compute index
+     * offsets for each removed element */
+    for (size_t i = JsonLength(parent); i > 0; i--)
+    {
+        size_t index = i - 1;
```

Comment:
Looks like this could easily turn into an off-by-one error, I'd consider being more explicit;

```suggestion
    for (size_t index_plus_one = JsonLength(parent); index_plus_one > 0; index_plus_one--)
    {
        size_t index = index_plus_one - 1;
```